bot/cogs/information.py

```diff
@@ -229,6 +233,81 @@ def __init__(self, bot: Bot):
 
         await ctx.send(embed=embed)
 
+    def format_fields(self, mapping: Mapping[str, Any], field_width: Optional[int] = None) -> str:
+        """Format a mapping to be readable to a human."""
+        # sorting is technically superfluous but nice if you want to look for a specific field
+        fields = sorted(mapping.items(), key=lambda item: item[0])
+
+        if field_width is None:
+            field_width = len(max(mapping.keys(), key=len))
+
+        out = ''
+
+        for key, val in fields:
+            if isinstance(val, dict):
+                # if we have dicts inside dicts we want to apply the same treatment to the inner dictionaries
+                inner_width = int(field_width * 1.6)
+                val = '\n' + self.format_fields(val, field_width=inner_width)
+
+            elif isinstance(val, str):
+                # split up text since it might be long
+                text = textwrap.fill(val, width=100, replace_whitespace=False)
+
+                # indent it, I guess you could do this with `wrap` and `join` but this is nicer
+                val = textwrap.indent(text, ' ' * (field_width + len(': ')))
+
+                # the first line is already indented so we `str.lstrip` it
+                val = val.lstrip()
+
+            if key == 'color':
+                # makes the base 10 representation of a hex number readable to humans
+                val = hex(val)
+
+            out += '{0:>{width}}: {1}\n'.format(key, val, width=field_width)
+
+        # remove trailing whitespace
+        return out.rstrip()
+
+    @group(invoke_without_command=True)
+    @in_channel(Channels.bot, bypass_roles=STAFF_ROLES)
+    async def raw(self, ctx: Context, *, message: discord.Message, json: bool = False) -> None:
+        """Shows information about the raw API response."""
+        # I *guess* it could be deleted right as the command is invoked but I felt like it wasn't worth handling
+        # doing this extra request is also much easier than trying to convert everything back into a dictionary again
+        raw_data = await ctx.bot.http.get_message(message.channel.id, message.id)
```

Comment:
can a user DOS the bot by spamming this command and making it get stuck in ratelimits?

Reply:
Yeah, most likely; this didn't occur to me. It should probably be gated behind a cooldown, and I don't know what the actual ratelimit is but some people said it's `1/1`. How strict should the cooldown be?

Reply:
very, IMO. I don't see a big need for regular users to run this on a constant basis

Reply:
but maybe the others have a better idea

Reply:
how about a strict cooldown for regular users and no cooldown for staff?

Reply:
that sounds like a good idea

Reply:
I've added a 2 per 3 minutes ratelimit with staff bypasses. I chose 2 because there is a subcommand that they might want to compare to the parent invokation and letting them do that immediately seemed not-too-bad. 3 minutes was arbitrary.